pkg/obfuscator/tracker.go

```diff
@@ -15,57 +15,66 @@ type ReplacementTracker interface {
-	AddReplacement(original string, replacement string)
+	AddReplacement(canonical, original string, replacement string)
+
+	// AddReplacementCount will add a replacement along with its original string to the report.
+	// Allows to set how many times 'original' occurs in source.
+	// If there is an existing value that does not match the given replacement, it will exit with a non-zero status.
+	AddReplacementCount(canonical, original string, replacement string, count uint)
 
 	// GenerateIfAbsent returns the previously used replacement if the entry is already present.
 	// If the replacement is not present then it uses the GenerateReplacement function to generate a replacement.
 	// The "key" parameter must be used for lookup and the "generator" parameter to generate the replacement.
-	GenerateIfAbsent(key string, generator GenerateReplacement) string
+	GenerateIfAbsent(canonical string, generator GenerateReplacement) string
 }
 
 type SimpleTracker struct {
 	lock    sync.RWMutex
-	mapping map[string]string
+	mapping map[string]Replacement
 }
 
-func (s *SimpleTracker) Report() map[string]string {
+var _ ReplacementTracker = (*SimpleTracker)(nil)
+
+func (s *SimpleTracker) Report() ReplacementReport {
 	s.lock.RLock()
 	defer s.lock.RUnlock()
-	defensiveCopy := make(map[string]string)
-	for k, v := range s.mapping {
-		defensiveCopy[k] = v
+	replacements := make([]Replacement, 0, len(s.mapping))
+	for _, v := range s.mapping {
+		replacements = append(replacements, v)
 	}
-	return defensiveCopy
+	return ReplacementReport{Replacements: replacements}
+}
+
+func (s *SimpleTracker) AddReplacement(canonical, original string, replacement string) {
+	s.AddReplacementCount(canonical, original, replacement, 1)
 }
 
-func (s *SimpleTracker) AddReplacement(original string, replacement string) {
+func (s *SimpleTracker) AddReplacementCount(canonical, original string, replacement string, count uint) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
-	if val, ok := s.mapping[original]; ok {
-		if replacement != val {
-			klog.Exitf("'%s' already has a value reported as '%s', tried to report '%s'", original, val, replacement)
-		}
+	if r, ok := s.mapping[canonical]; ok {
+		s.mapping[canonical] = r.Increment(original, count)
 		return
 	}
-	s.mapping[original] = replacement
+	new := Replacement{Canonical: canonical, ReplacedWith: replacement}
+	new.Increment(original, count)
+	s.mapping[canonical] = new
 }
 
-func (s *SimpleTracker) GenerateIfAbsent(key string, generator GenerateReplacement) string {
+func (s *SimpleTracker) GenerateIfAbsent(canonical string, generator GenerateReplacement) string {
 	s.lock.Lock()
 	defer s.lock.Unlock()
-	if val, ok := s.mapping[key]; ok {
-		return val
+	if r, ok := s.mapping[canonical]; ok {
+		return r.ReplacedWith
 	}
 	if generator == nil {
 		return ""
 	}
-	r := generator()
-	s.mapping[key] = r
-	return r
+	return generator()
 }
 
 func (s *SimpleTracker) Initialize(replacements map[string]string) {
```

Comment:
I'm not sure whether this is currently used, probably not. But that ideally should take your `ReplacementReport`. I assume that also is touching @sairameshv a bit  - fyi.